test/apps/iodemo/ucx_wrapper.cc

```diff
@@ -643,6 +659,49 @@ void UcxConnection::accept(ucp_conn_request_h conn_req, UcxCallback *callback)
     connect_common(ep_params, callback);
 }
 
+void UcxConnection::disconnect(UcxCallback *callback)
+{
+    /* establish cb must be destroyed earlier since it accesses
+     * the connection */
+    assert(_establish_cb == NULL);
+    assert(_disconnect_cb == NULL);
+    assert(_ep != NULL);
+
+    UCX_CONN_LOG << "destroying, ep is " << _ep;
+
+    _disconnect_cb = callback;
+    ep_close(UCP_EP_CLOSE_MODE_FORCE);
+
+    if (ucs_list_is_empty(&_all_requests)) {
+        _context.remove_connection(this);
+        _context._disconnecting_conns.push_back(this);
+    } else {
+        cancel_all();
+    }
+}
+
+bool UcxConnection::disconnect_progress()
+{
+    assert(_ep == NULL);
+    assert(_disconnect_cb != NULL);
+
+    if (UCS_PTR_IS_PTR(_close_request)) {
+        if (ucp_request_check_status(_close_request) == UCS_INPROGRESS) {
+            return false;
+        } else {
+            ucp_request_free(_close_request);
+            _close_request = NULL;
+        }
+    }
+
+    assert(ucs_list_is_empty(&_all_requests));
```

Comment:
maybe better move it to UcxConnection destructor?

Reply:
do you mean `assert(ucs_list_is_empty(&_all_requests));`? it's already there and should be true in both places

Reply:
so why need it here as well? IMO, dtor is enough

Reply:
because we have to move connection from _conns contained to disconnecting list when all requests are completed